pkg/kubelet/status/state/state_checkpoint_test.go

```diff
@@ -0,0 +1,158 @@
+			}
+			tests = append(tests, struct {
+				name string
+				args args
+			}{
+				name: fmt.Sprintf("resource - %s%s", fact, suf),
+				args: args{
+					podResourceAllocation: PodResourceAllocation{
+						"pod1": {
+							"container1": {
+								v1.ResourceCPU:    resource.MustParse(fmt.Sprintf("%s%s", fact, suf)),
+								v1.ResourceMemory: resource.MustParse(fmt.Sprintf("%s%s", fact, suf)),
+							},
+						},
+					},
+				},
+			})
+		}
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := sc.cache.ClearState()
+			require.NoError(t, err, "failed to clear state")
+
+			defer func() {
+				err = sc.checkpointManager.RemoveCheckpoint(sc.checkpointName)
+				require.NoError(t, err, "failed to remove checkpoint")
+			}()
+
+			err = sc.cache.SetPodResourceAllocation(tt.args.podResourceAllocation)
+			require.NoError(t, err, "failed to set pod resource allocation")
+
+			err = sc.storeState()
+			require.NoError(t, err, "failed to store state")
+
+			// deep copy cache
+			originCache := NewStateMemory()
+			podAllocation := sc.cache.GetPodResourceAllocation()
+			err = originCache.SetPodResourceAllocation(podAllocation)
+			require.NoError(t, err, "failed to set pod resource allocation")
+
+			err = sc.cache.ClearState()
+			require.NoError(t, err, "failed to clear state")
+
+			err = sc.restoreState()
+			require.NoError(t, err, "failed to restore state")
+
+			restoredCache := sc.cache
+			require.Equal(t, len(originCache.GetPodResourceAllocation()), len(restoredCache.GetPodResourceAllocation()), "restored pod resource allocation is not equal to original pod resource allocation")
+			for podUID, containerResourceList := range originCache.GetPodResourceAllocation() {
+				require.Equal(t, len(containerResourceList), len(restoredCache.GetPodResourceAllocation()[podUID]), "restored pod resource allocation is not equal to original pod resource allocation")
+				for containerName, resourceList := range containerResourceList {
+					for name, quantity := range resourceList {
+						if !quantity.Equal(restoredCache.GetPodResourceAllocation()[podUID][containerName][name]) {
+							t.Errorf("restored pod resource allocation is not equal to original pod resource allocation")
+						}
+					}
+				}
+			}
+		})
+	}
+}
+
+func Test_stateCheckpoint_formatUpgraded(t *testing.T) {
+	// Based on the PodResourceAllocationInfo struct, it's mostly possible that new field will be added
+	// in struct PodResourceAllocationInfo, rather than in struct PodResourceAllocationInfo.AllocationEntries.
+	// Emulate upgrade scenario by pretending that `ResizeStatusEntries` is a new field.
+	// The checkpoint content doesn't have it and that shouldn't prevent the checkpoint from being loaded.
+	checkpointContent := `{"data":"{\"allocationEntries\":{\"pod1\":{\"container1\":{\"cpu\":\"1Ki\",\"memory\":\"1Ki\"}}}}","checksum":143149502}`
+	checkpoint := &Checkpoint{}
+	err := checkpoint.UnmarshalCheckpoint([]byte(checkpointContent))
```

Comment:
Can you use stateCheckpoint APIs here to cover a bit more functionality?

Reply:
Updated

Reply:
Thanks. BTW, my suggestions to add checksum verification to the UnmarshallCheckpoint call were caused by this line as it was simply unmarshalling content without verification.